src/Microsoft.VisualStudio.ProjectSystem.Managed/PooledObjects/PooledArray.cs

```diff
@@ -0,0 +1,437 @@
+            }
+
+            var tmp = PooledArray<U>.GetInstance(Count);
+            foreach (T i in this)
+            {
+                tmp.Add((U)i);
+            }
+
+            return tmp.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Realizes the array and disposes the builder in one operation.
+        /// </summary>
+        public ImmutableArray<T> ToImmutableAndFree()
+        {
+            ImmutableArray<T> result;
+            if (_builder.Capacity == Count)
+            {
+                result = _builder.MoveToImmutable();
+            }
+            else
+            {
+                result = ToImmutable();
+            }
+
+            Free();
+            return result;
+        }
+
+        public T[] ToArrayAndFree()
+        {
+            T[] result = ToArray();
+            Free();
+            return result;
+        }
+
+        // To implement Poolable, you need two things:
+        // 1) Expose Freeing primitive. 
+        public void Free()
+        {
+            ObjectPool<PooledArray<T>> pool = _pool;
+            if (pool != null)
+            {
+                // We do not want to retain (potentially indefinitely) very large builders 
+                // while the chance that we will need their size is diminishingly small.
+                // It makes sense to constrain the size to some "not too small" number. 
+                // Overall perf does not seem to be very sensitive to this number, so I picked 128 as a limit.
+                if (_builder.Capacity < 128)
+                {
+                    if (Count != 0)
+                    {
+                        Clear();
+                    }
+
+                    pool.Free(this);
+                    return;
+                }
+            }
+        }
+
+        // 2) Expose the pool or the way to create a pool or the way to get an instance.
+        //    for now we will expose both and figure which way works better
+        private static readonly ObjectPool<PooledArray<T>> s_poolInstance = CreatePool();
+        public static PooledArray<T> GetInstance()
+        {
+            PooledArray<T> builder = s_poolInstance.Allocate();
+            return builder;
+        }
+
+        public static PooledArray<T> GetInstance(int capacity)
+        {
+            PooledArray<T> builder = GetInstance();
+            builder.EnsureCapacity(capacity);
+            return builder;
+        }
+
+        public static PooledArray<T> GetInstance(int capacity, T fillWithValue)
+        {
+            PooledArray<T> builder = GetInstance();
+            builder.EnsureCapacity(capacity);
+
+            for (int i = 0; i < capacity; i++)
+            {
+                builder.Add(fillWithValue);
+            }
+
+            return builder;
+        }
+
+        public static ObjectPool<PooledArray<T>> CreatePool()
+        {
+            return CreatePool(128); // we rarely need more than 10
```

Comment:
This line confuses me.

Reply:
`128` is the size of the array of objects, however in practice (for roslyn where this was stolen) it rarely allocates 128 objects.  most of the objects in the array are null.

Reply:
I was mostly referring to the comment.

Reply:
...which I see you've just fixed. 👍